examples/step-77/step-77.cc

```diff
@@ -0,0 +1,663 @@
+    for (types::global_dof_index i :
+         DoFTools::extract_hanging_node_dofs(dof_handler))
+      residual(i) = 0;
+
+    std::cout << " norm=" << residual.l2_norm() << std::endl;
+  }
+
+
+
+  // @sect4{Solving linear systems with the Jacobian matrix}
+
+  // Next up is the function that implements the solution of a linear system
+  // with the Jacobian matrix. Since we have already factored the matrix when we
+  // built the matrix, solving a linear system comes down to applying the
+  // inverse matrix to the given right hand side vector: This is what the
+  // SparseDirectUMFPACK::vmult() function does that we use here. Following
+  // this, we have to make sure that we also address the values of hanging nodes
+  // in the solution vector, and this is done using
+  // AffineConstraints::distribute().
+  //
+  // The function takes an addition, but unused, argument `tolerance` that
+  // indicates how accurately we have to solve the linear system. The meaning of
+  // this argument is discussed in the introduction in the context of the
+  // "Eisenstat Walker trick", but since we are using a direct rather than an
+  // iterative solver, we are not using this opportunity to solve linear systems
+  // only inexactly.
+  template <int dim>
+  void MinimalSurfaceProblem<dim>::solve(const Vector<double> &rhs,
+                                         Vector<double> &      solution,
+                                         const double /*tolerance*/)
+  {
+    TimerOutput::Scope t(computing_timer, "linear system solve");
+
+    std::cout << "  Solving linear system" << std::endl;
+
+    jacobian_matrix_factorization->vmult(solution, rhs);
+
+    hanging_node_constraints.distribute(solution);
+  }
+
+
+
+  // @sect4{Refining the mesh, setting boundary values, and generating graphical output}
+
+  // The following three functions are again simply copies of the ones in
+  // step-15:
+  template <int dim>
+  void MinimalSurfaceProblem<dim>::refine_mesh()
+  {
+    Vector<float> estimated_error_per_cell(triangulation.n_active_cells());
+
+    KellyErrorEstimator<dim>::estimate(
+      dof_handler,
+      QGauss<dim - 1>(fe.degree + 1),
+      std::map<types::boundary_id, const Function<dim> *>(),
+      current_solution,
+      estimated_error_per_cell);
+
+    GridRefinement::refine_and_coarsen_fixed_number(triangulation,
+                                                    estimated_error_per_cell,
+                                                    0.3,
+                                                    0.03);
+
+    triangulation.prepare_coarsening_and_refinement();
+
+    SolutionTransfer<dim> solution_transfer(dof_handler);
+    solution_transfer.prepare_for_coarsening_and_refinement(current_solution);
+
+    triangulation.execute_coarsening_and_refinement();
+
+    dof_handler.distribute_dofs(fe);
+
+    Vector<double> tmp(dof_handler.n_dofs());
+    solution_transfer.interpolate(current_solution, tmp);
+    current_solution = tmp;
```

Comment:
Does `std::swap` work on `Vector`?

Reply:
It would, but it seems harder to read. I'm using `current_solution = std::move(tmp)` instead now.